google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/strategy/CreateNodesWithGroupEncodedIntoNameThenAddToSet.java

```diff
@@ -134,52 +128,51 @@ protected CreateNodesWithGroupEncodedIntoNameThenAddToSet(
-         String name = naming.name(port);
-         Firewall firewall = firewallApi.get(name);
-         if (firewall == null) {
-            ImmutableSet<Firewall.Rule> rules = ImmutableSet.of(Firewall.Rule.permitTcpRule(port), Firewall.Rule.permitUdpRule(port));
-            FirewallOptions firewallOptions = new FirewallOptions()
-                    .name(name)
-                    .network(network.getSelfLink())
-                    .allowedRules(rules)
-                    .sourceTags(templateOptions.getTags())
-                    .sourceRanges(of(DEFAULT_INTERNAL_NETWORK_RANGE, EXTERIOR_RANGE))
-                    .targetTags(ImmutableSet.of(name));
-            AtomicReference<Operation> operation = Atomics.newReference(firewallApi.createInNetwork(
-                    firewallOptions.getName(),
-                    network.getSelfLink(),
-                    firewallOptions));
-            operations.add(operation);
-         }
-      }
-
-      for (AtomicReference<Operation> operation : operations) {
+      Firewall firewall = firewallApi.get(firewallName);
+      Set<Rule> rules = createFirewallRulesFromInboundPorts(templateOptions.getInboundPorts());
+      FirewallOptions firewallOptions = new FirewallOptions().name(firewallName).network(network.getSelfLink())
+                                                             .sourceTags(templateOptions.getTags())
+                                                             .sourceRanges(of(DEFAULT_INTERNAL_NETWORK_RANGE,
+                                                                     EXTERIOR_RANGE));
+      if (firewall == null) {
+         firewallOptions.allowedRules(rules);
+         AtomicReference<Operation> operation = newReference(firewallApi.createInNetwork(firewallOptions.getName(),
+                 network.getSelfLink(),
+                 firewallOptions));
+         retry(operationDonePredicate, operationCompleteCheckTimeout, operationCompleteCheckInterval,
+                 MILLISECONDS).apply(operation);
+         checkState(!operation.get().getHttpError().isPresent(), "Could not create firewall, operation failed " + operation);
+      } else {
+         rules.addAll(firewall.getAllowed());
+         AtomicReference<Operation> operation = newReference(firewallApi.patch(firewallOptions.getName(), firewallOptions));
          retry(operationDonePredicate, operationCompleteCheckTimeout, operationCompleteCheckInterval,
                  MILLISECONDS).apply(operation);
-         checkState(!operation.get().getHttpError().isPresent(),"Could not create firewall, operation failed" + operation);
+         checkState(!operation.get().getHttpError().isPresent(), "Could not patch firewall, operation failed " + operation);
+      }
+   }
+
+   private Set<Rule> createFirewallRulesFromInboundPorts(int[] inboundPorts) {
+      Set<Rule> rules = Sets.newLinkedHashSet();
+      for (int port : inboundPorts) {
+         rules.add(Rule.permitTcpRule(port));
```

Comment:
This shouldn't add one rule per port. Rules can specify arrays of ranges. We need some logic to find the minimal number of rules for the given inbound ports. We should specify one rule for Tcp and one rule for Udp and each of those rules should enumerate all the ports.

using the ports form this pastebin as an example: http://pastebin.com/3WxHWj2X
[22, 4369, 6000-7999, 8087, 8093, 8098-8099, 8985] covers all the ports
```
"allowed": [
  {
    "IPProtocol": "tcp",
    "ports":["22", "4369", "6000-7999", "8087", "8093", "8098-8099", "8985"],
  },
  {
    "IPProtocol": "udp",
    "ports": ["22", "4369", "6000-7999", "8087", "8093", "8098-8099", "8985"] ,
  }
]
```

https://cloud.google.com/compute/docs/networking#firewalls_1 is a good resource